File: apps/spark/java/livy-server/src/main/java/com/cloudera/hue/livy/server/sessions/Statement.java
package com.cloudera.hue.livy.server.sessions;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.ArrayList;
import java.util.List;

public class Statement {

    public enum State {
        NOT_READY,
        READY,
        INCOMPLETE,
        RUNNING,
        COMPLETE,
    }

    int id;
    State state;
    final List<String> input = new ArrayList<String>();
    final List<String> output = new ArrayList<String>();

    final List<String> error = new ArrayList<String>();

    public Statement(int id) {
        this.id = id;
        this.state = State.COMPLETE;
    }

    @JsonProperty
    public int getId() {
        return id;
    }

    @JsonProperty("state")
    public State getState() {
        return state;
    }

    public void setState(State state) {
        this.state = state;
    }

    @JsonProperty("input")
    public List<String> getInput() {
        return input;
    }

    public void addInput(String input) {
        this.input.add(input);
    }

    @JsonProperty("output")
    public List<String> getOutput() {
        return output;
    }

    public void addOutput(String output) {
        this.output.add(output);
    }

    @JsonProperty("error")
    public List<String> getError() {
        return error;
    }

    public void addError(String error) {
        this.error.add(error);
    }
}

File: apps/spark/java/livy-server/src/main/scala/com/cloudera/hue/livy/server/sessions/Statement.scala
package com.cloudera.hue.livy.server.sessions

import com.cloudera.hue.livy.msgs.ExecuteRequest
import org.json4s.JValue

import scala.concurrent.{ExecutionContext, ExecutionContextExecutor, Future}
import scala.util.{Failure, Success}

object Statement {
  sealed trait State

  case class Running() extends State {
    override def toString = "running"
  }

  case class Available() extends State {
    override def toString = "available"
  }

  case class Error() extends State {
    override def toString = "error"
  }
}

class Statement(val id: Int, val request: ExecuteRequest, val output: Future[JValue]) {
  import Statement._

  protected implicit def executor: ExecutionContextExecutor = ExecutionContext.global

  private[this] var _state: State = Running()

  def state = _state

  output.onComplete {
    case Success(_) => _state = Available()
    case Failure(_) => _state = Error()
  }
}
